src/command/restore/timeline.c

```diff
@@ -0,0 +1,293 @@
+                }
+                TRY_END();
+            }
+
+            // Only proceed if target timeline is not the backup timeline
+            if (timelineTarget != timelineBackup)
+            {
+                // Search through the history for the target timeline to make sure it includes the backup timeline. This follows
+                // the logic in PostgreSQL's readTimeLineHistory() and tliOfPointInHistory() but is a bit simplified for our usage.
+                const List *const historyList = historyLoad(storageRepo, archiveId, timelineTarget, cipherType, cipherPass);
+                unsigned int timelineFound = 0;
+
+                for (unsigned int historyIdx = 0; historyIdx < lstSize(historyList); historyIdx++)
+                {
+                    const HistoryItem *const historyItem = lstGet(historyList, historyIdx);
+
+                    // If backup timeline exists in the target timeline's history before the fork LSN then restore can proceed
+                    if (historyItem->timeline == timelineBackup && lsnBackup < historyItem->lsn)
+                    {
+                        timelineFound = timelineBackup;
+                        break;
+                    }
+                }
+
+                // Error when the timeline was not found or not found in the expected range
+                if (timelineFound == 0)
+                {
+                    uint64_t lsnFound = 0;
+
+                    // Check if timeline exists but did not fork off when expected
+                    for (unsigned int historyIdx = 0; historyIdx < lstSize(historyList); historyIdx++)
+                    {
+                        const HistoryItem *const historyItem = lstGet(historyList, historyIdx);
+
+                        if (historyItem->timeline == timelineBackup)
+                        {
+                            timelineFound = timelineBackup;
+                            lsnFound = historyItem->lsn;
+                        }
+                    }
+
+                    // The timeline was found but it forked off earlier than the backup LSN. This covers the common case where a
+                    // standby is promoted by accident, which creates a new timeline, and then the user tries to restore a backup
+                    // from the original primary that was made after the new timeline was created. Since PostgreSQL >= 12 will
+                    // always try to follow the latest timeline this restore will fail since there is no path to the target
+                    // timeline.
+                    if (timelineFound != 0)
+                    {
+                        ASSERT(lsnFound != 0);
+
+                        THROW_FMT(
+                            DbMismatchError,
+                            "target timeline %X forked from backup timeline %X at %s which is before backup lsn of %s\n"
+                            "HINT: was the target timeline created by accidentally promoting a standby?\n"
```

Comment:
This situation can probably even pretty easily with test restores too. Could we add `or performing restore tests` to the hint maybe?

Reply:
Good idea. How about a6ff644071d89e.

Reply:
Yep, sounds good indeed, that will remind the users one of our most frequent recommendation :+1: